// yanl/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function createColorFn(color) {
    var open = "\u001B[" + color + "m";
    var close = '\u001b[39m';
    return function (value) { return "" + open + value + close; };
}
exports.colors = {
    cyan: createColorFn(95),
    gray: createColorFn(90),
};
function padZero(value, len) {
    if (len === void 0) { len = 2; }
    return value.toString().padStart(len, '0');
}
function timestamp() {
    var date = new Date();
    var hours = date.getHours();
    var minutes = date.getMinutes();
    var seconds = date.getSeconds();
    return [hours, minutes, seconds].map(padZero).join(':');
}
exports.timestamp = timestamp;
function asJSON(value) {
    try {
        return JSON.parse(JSON.stringify(value));
    }
    catch (_a) {
        return value;
    }
}
exports.asJSON = asJSON;
function getCaller() {
    var _a, _b, _c;
    var levels = 4;
    var trace = new Error().stack;
    var caller = (_a = trace) === null || _a === void 0 ? void 0 : _a.split('\n').slice(levels, levels + 1);
    return (_c = (_b = caller) === null || _b === void 0 ? void 0 : _b[0].match(/at (.+)\s/)) === null || _c === void 0 ? void 0 : _c[1];
}
exports.getCaller = getCaller;
